aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -291,12 +291,12 @@ Tensor prod(const Tensor &self) {
 static inline Tensor &mean_out(Tensor &result, const Tensor &self, IntArrayRef dim,
                  bool keepdim, optional<ScalarType> opt_dtype) {
   ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
-  TORCH_CHECK(
-      at::isFloatingType(scalarType),
-      "Can only calculate the mean of floating types. Got ",
-      toString(scalarType),
-      " instead.");
   ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+
+  if ((at::isIntegralType(scalarType) || scalarType == ScalarType::Bool) &&
+      !opt_dtype.has_value()) {
+    dtype = c10::typeMetaToScalarType(c10::get_default_dtype());
```

Comment:
If I'm understanding this correctly, we are promoting integral dtype to whatever the default dtype is when one does `torch.mean`. I am not sure this is the behavior we want.

cc @gchanan do we have a plan for type promotion?

Reply:
I think this is distinct from type promotion -- i.e. this never needs to read from the promotion table, and you can tell this because the type of the input does not affect the output.

This sounds like the correct behavior.